pkg/machines/hostvmslist.jsx

```diff
@@ -463,41 +467,93 @@ Vm.propTypes = {
+    constructor(props) {
+        super(props);
+        this.deviceProxyHandler = this.deviceProxyHandler.bind(this);
+        this.deviceProxyAddHandler = this.deviceProxyAddHandler.bind(this);
+        this.deviceProxyRemoveHandler = this.deviceProxyRemoveHandler.bind(this);
+        this.state = { hostDevices: []};
+        this.client = cockpit.dbus("org.freedesktop.NetworkManager", {});
     }
 
-    const sortFunction = (vmA, vmB) => vmA.name.localeCompare(vmB.name);
+    componentWillMount() {
+        const deviceProxies = this.client.proxies("org.freedesktop.NetworkManager.Device");
+        deviceProxies.addEventListener('added', this.deviceProxyAddHandler);
+        deviceProxies.addEventListener('changed', this.deviceProxyHandler);
+        deviceProxies.addEventListener('removed', this.deviceProxyRemoveHandler);
+    }
 
-    let allActions = []; // like createVmAction
-    if (actions) {
-        allActions = allActions.concat(actions);
+    componentWillUnmount() {
+        this.client.close();
     }
 
-    return (<div className='container-fluid'>
-        <Listing title={_("Virtual Machines")} columnTitles={[_("Name"), _("Connection"), _("State")]} actions={allActions}>
-            {vms
-                .sort(sortFunction)
-                .map(vm => {
-                return (
-                    <Vm vm={vm} config={config}
-                        onStart={() => dispatch(startVm(vm))}
-                        onReboot={() => dispatch(rebootVm(vm))}
-                        onForceReboot={() => dispatch(forceRebootVm(vm))}
-                        onShutdown={() => dispatch(shutdownVm(vm))}
-                        onForceoff={() => dispatch(forceVmOff(vm))}
-                        onUsageStartPolling={() => dispatch(usageStartPolling(vm))}
-                        onUsageStopPolling={() => dispatch(usageStopPolling(vm))}
-                        onSendNMI={() => dispatch(sendNMI(vm))}
-                        dispatch={dispatch}
-                    />);
-            })}
-        </Listing>
-    </div>);
-};
+    deviceProxyAddHandler(event, proxy) {
+        const devices = this.state.hostDevices;
+        if (devices.indexOf(proxy.Interface) === -1) {
+            devices.push(proxy.Interface);
+        }
+        this.setState({ hostDevices: devices });
+    }
+
+    deviceProxyHandler(event, proxy) {
+        const devices = this.state.hostDevices;
+        const index = devices.indexOf(proxy.Interface);
+        if (index > -1 && !proxy.valid) {
+            devices.splice(index, 1);
+        }
+        this.setState({ hostDevices: devices });
+    }
+
+    deviceProxyRemoveHandler(event, proxy) {
+        const devices = this.state.hostDevices;
+        const index = devices.indexOf(proxy.Interface);
+        if (index > -1) {
+            devices.splice(index, 1);
+        }
+        this.setState({ hostDevices: devices });
+    }
+
+    render() {
+        const that = this;
```

Comment:
This kind of dance is not really necessary with ES6 any more. `that` below should be exactly the same as `this` as you only use arrow functions, not "full" `function()`s.